Scaladoc: Refactor static site loading and directory structure

pikinier20 opened the pull request:
Static site in Scaladoc is probably one of its core features. Since it didn't have a lot of attention and also due to changing requirements, static site used to be not-so-clear piece of code. Having in mind upcoming changes to the Scaladoc, I decided to do a bigger refactor mainly in static site loading phase and also in directory structure.

I think that before merging we should also update documentation.

# Static site loading

Previously, we loaded all files from filesystem and the `sidebar.yml` file was only used to generate LHS navigation. This solution was a bit problematic. Firstly, we could have pages that were only navigable by other pages or by sharing a link. It doesn't seem as a good thing because we could easily loose control over what's in the navigation and what is rendered. Also, it caused a problem with index pages: every section in navigation needs to have a index page otherwise we would get 404 after entering a section. Since the `sidebar.yml` definition had nothing to do with location of files in filesystem, there was a problem with finding a location where the index should be rendered. (Navigation could e.g. aggregate files among more than one directory). This solution is also not flexible as we can't change the place where the file is rendered (Rendered directory structure is always the same as the source directory structure).

I've changed that so we have two ways of loading static site:
1. We don't create `sidebar.yml` file and Scaladoc will load all documents that are present in <site-root>/_docs/ directory and automatically generate navigation.
2. We create `sidebar.yml` file (probably we also need to rename it?) which declares sections and pages that should be rendered and navigable. (We can also add a option: hidden to render without entry in navigation). The main advantage though is that the YAML pages structure defines the rendered directory structure. By default, for each section there will be directory created by converting its name to kebab case (e.g. Other New Features => other-new-features/). There's a setting `directory` that overrides this convention. We can also define a section with title and index but without defining children which will cause loading documents from the directory of index file. e.g.
```
- title: Some section
  index: some-section/index.md
```
will load all files from `some-section/` directory.

# Static site directory structure

After my changes, static site directory structure looks like this:
```
.
└── <site-root>/
    ├── _docs/
    │   └── ...
    ├── _blog/
    │   ├── index.md
    │   └── _posts/
    │       └── ...
    └── _assets/
        ├── js/
        │   └── ...
        ├── img/
        │   └── ...
        └── ...
```
All documents should be in `_docs` directory. File-system loading crawls this directory and also every entry in YAML configuration is relative to this directory.

Whole blog content should be in `_blog` directory. We check if the directory exists and based on that, we generate Blog section or not.

All assets should be in `_assets` directory. The whole content of this directory is copied to the documentation and there's a check if asset exists by checking if the given path resolved in `_assets` exists. e.g. for `images/logo.png` we check if `_assets/images/logo.png` exists.

This change makes a static site structure clearer and also simplifies a lot of logic in codebase.

# The `-Yapi-subdirectory` setting

The setting was introduced to resolve possible conflicts between static site and API. However, we also decided to use it in order to define what is a top-level documentation. With the `-Yapi-subdirectory` off (default), API documentation is in top-level and whole static site is inside `_docs` subdirectory. Otherwise, the static side goes to top-level and API is in `api` subdirectory.

BarkingBad commented on scaladoc/src/dotty/tools/scaladoc/renderers/Locations.scala:
```diff
@@ -34,9 +34,9 @@ trait Locations(using ctx: DocContext):
       case null =>
         val path = dri match
-          case `docsRootDRI` => List("docs", "index")
+          // case `docsRootDRI` => List("docs", "index")
```
Also leftover

BarkingBad commented on scaladoc/src/dotty/tools/scaladoc/renderers/Renderer.scala:
```diff
@@ -82,34 +74,17 @@ abstract class Renderer(rootPackage: Member, val members: Map[DRI, Member], prot
           updatedTemplates.result()
 
-        val newTemplates = updateSettings(templates, newSettings.to(ListBuffer))
+        val newTemplates = updateSettings(Seq(siteContext.staticSiteRoot.rootTemplate), newSettings.to(ListBuffer))
```
I guess it was supposed to be `Seq(rootTemplate)` or wh do we need property in line 43?

BarkingBad commented on scaladoc/src/dotty/tools/scaladoc/site/StaticSiteContext.scala:
```diff
@@ -19,63 +20,30 @@ class StaticSiteContext(
-    if files.size > 1 then
-      val msg = s"ERROR: Multiple root index pages found: ${files.map(_.getAbsolutePath)}"
-      report.error(msg)
+  val relativizeFrom = if args.apiSubdirectory then docsPath else root.toPath
```
Nice trick :D

BarkingBad commented on scaladoc/src/dotty/tools/scaladoc/site/StaticSiteLoader.scala:
```diff
@@ -0,0 +1,201 @@
+        LoadedTemplate(withUpdatedTitle, List.empty, pathFromRoot.resolve(file.getName).toFile)
+      case Sidebar.Root(_, _) =>
+        // Cannot happen
+        ???
```
Maybe we could log it instead of throwing `NotImplementedException` even though it will never hit this branch

BarkingBad commented on scaladoc/src/dotty/tools/scaladoc/site/templates.scala:
```diff
@@ -105,7 +115,25 @@ case class TemplateFile(
+          case "diff" =>
+            s"<pre><code class=\"language-diff hljs\" data-lang=\"diff\">${super.asString(nodes(1).render(context), context)}</code></pre>\n\n"
+          case _ =>
+            report.warn("Unsupported highlight value. Currenlty supported values are: `diff`", file)(using ssctx.outerCtx)
```
`Currently` typo

julienrf reviewed (commented):
Nice work @pikinier20!

There are several parts that I don’t understand in the diff, I have left some comments, do you mind clarifying?

Other than that, here are a few comments about the PR description:

> I think that before merging we should also update documentation.

Sure! Thank you!

> 1. We don't create sidebar.yml file and Scaladoc will load all documents that are present in /_docs/ directory and automatically generate navigation.

So, the navigation bar will mirror the hierarchy of directories and files in `_docs`? What would be the order of the files? Alphabetical, I guess?

> 2. We create sidebar.yml file (probably we also need to rename it?)

Why do you want to rename it?

> (We can also add a option: hidden to render without entry in navigation)

What do you mean? I don’t understand this idea.

> There's a setting `directory` that overrides this convention.

Could you please show an example?

> With the -Yapi-subdirectory off (default), API documentation is in top-level and whole static site is inside _docs subdirectory. Otherwise, the static side goes to top-level and API is in api subdirectory.

Did you make any changes to `-Yapi-subdirectory` in this PR? I can’t see them.

pikinier20 commented on scaladoc-testcases/src/_docs/tests/Adoc.scala:
```diff
@@ -1,4 +1,4 @@
-package docs.tests
+package _docs.tests
```
Yes, because it's used in the test checking conflicts between static site and api

pikinier20 commented on scaladoc/src/dotty/tools/scaladoc/renderers/HtmlRenderer.scala:
```diff
@@ -31,32 +31,23 @@ class HtmlRenderer(rootPackage: Member, members: Map[DRI, Member])(using ctx: Do
+          Using(Files.walk(file)) { stream =>
+            stream.iterator().asScala.toSeq
+              .map(from => Resource.File(resourceFile.toPath.relativize(from).toString, from))
+          }.get
```
I don't know if you selected proper diff for this message?

julienrf replied:
IIUC, we call `get` on a `Try` value, which will throw an exception in case of failure. We should make sure the error message is meaningful (actionable for the end-user), and if not, we should report a better error message.

pikinier20 commented on scaladoc/src/dotty/tools/scaladoc/site/SidebarParser.scala:
```diff
@@ -8,29 +8,46 @@ import collection.JavaConverters._
+
+  def load(content: String): Sidebar.Root =
+    val mapper = ObjectMapper(YAMLFactory())
+    val root: RawRoot = mapper.readValue(content, RootTypeRef)
```
Yes, probably we need to handle exceptions thrown by jackson library and report the issue to the user

pikinier20 commented on scaladoc/src/dotty/tools/scaladoc/site/StaticSiteLoader.scala:
```diff
@@ -0,0 +1,201 @@
+      }
+  }
+
+  def loadBasedOnYaml(yamlRoot: Sidebar.Root): StaticSiteRoot = {
```
Sure

pikinier20 commented on scaladoc/src/dotty/tools/scaladoc/site/templates.scala:
```diff
@@ -105,7 +115,25 @@ case class TemplateFile(
     val mutableProperties = new JHashMap(ctx.properties.transform((_, v) => asJavaElement(v)).asJava)
 
-    val rendered = Template.parse(this.rawCode).render(mutableProperties)
+    val tag = new Tag("highlight"):
```
For the Scala 3 website only

pikinier20 commented on scaladoc/src/dotty/tools/scaladoc/site/templates.scala:
````diff
@@ -105,7 +115,25 @@ case class TemplateFile(
+            s"```${super.asString(nodes(1).render(context), context)}```\n\n"
+
+    val tag2 = new Tag("link"):
+      override def render(context: TemplateContext, nodes: Array[? <: LNode]): Object =
+        super.asString(nodes(0).render(context), context) match
+          case sth =>
+            report.warn(s"Unsupported link tag. Link to $sth can't be resolved", file)(using ssctx.outerCtx)
+            "/"
````
It is for Scala 3 website purpose only and I don't really know how to handle this tag. Maybe we should just manually remove it from markdowns and not handle it at all.

julienrf replied:
Which pages use it?

pikinier20 commented on scaladoc/test-documentations/basic/index.md:
```diff
@@ -1,8 +0,0 @@
-
-[Link to docs](docs/index.html)
-
-And a text!
```
The file is removed because it's not used anymore - there was kind of mess with indexes: we were able to define docs index, api index and landing page. With my changes, the landing page is always either api index or docs index depending on the `-Yapi-subdirectory` setting. The reason is that I think the root index of the documentation should be a doc page already. Custom landing pages are not necessarily a responsibility of doctool.

pikinier20 commented on docs/index.html:
```diff
@@ -1,109 +0,0 @@
----
```
This is a thing that we need to discuss. As I said in comment above - I don't think that custom landing pages are a responsibility of Scaladoc. For me, the entry point to the generated documentation should be standard doc page with the navigation etc. If we want to keep the "red" landing page on dotty.epfl.ch maybe we should just deploy it in the root directory and the entire generated nightly documentation would land into separate directory e.g. "reference" or "documentation"?

pikinier20 replied:
I don't really see any big advantage of generating custom landing page by Scaladoc. One can always do it manually. Usually, the links on landing page are not changed very often. Also, if we want to integrate Scaladoc and Jekyll, we can generate this page with Jekyll.

pikinier20 commented:
> Nice work @pikinier20!
> 
> There are several parts that I don’t understand in the diff, I have left some comments, do you mind clarifying?
> 
> Other than that, here are a few comments about the PR description:
> 
> > I think that before merging we should also update documentation.
> 
> Sure! Thank you!
> 
> > 1. We don't create sidebar.yml file and Scaladoc will load all documents that are present in /_docs/ directory and automatically generate navigation.
> 
> So, the navigation bar will mirror the hierarchy of directories and files in `_docs`? What would be the order of the files? Alphabetical, I guess?

Alphabetical sounds good for me.
> 
> > 2. We create sidebar.yml file (probably we also need to rename it?)
> 
> Why do you want to rename it?
> 
I just want to support multiple filenames because `sidebar` associates with just a navigation and in this case it decides what lands in static site or not.
> > (We can also add a option: hidden to render without entry in navigation)
> 
> What do you mean? I don’t understand this idea.
> 
The `hidden` option in loading using `YAML` would be meant for pages that we want to have in static site but we don't want to have an entry in navigation for them. (If the page is not declared in YAML, it won't be rendered)
> > There's a setting `directory` that overrides this convention.
> 
> Could you please show an example?
> 
Loading using YAML by default converts section title to the directory name that will contain the section. So for example:
```yaml
- title: Some section
  subsection:
    ...
```
would be rendered inside `some-section/` directory. Sometimes we want to override the directory name. To do so, we can define section as:
```yaml
- title: Some section
  directory: custom-directory
  subsection:
    ...
```
which results in rendering this section in `custom-directory/` directory.
> > With the -Yapi-subdirectory off (default), API documentation is in top-level and whole static site is inside _docs subdirectory. Otherwise, the static side goes to top-level and API is in api subdirectory.
> 
> Did you make any changes to `-Yapi-subdirectory` in this PR? I can’t see them.

There's an added line in `StaticSiteContext` which decides if the docs index lands into root directory or in the _docs subdirectory. Same logic is added in `Locations` which does the same for API.

julienrf commented:
> I just want to support multiple filenames because `sidebar` associates with just navigation and in this case it decides what lands in static site or not.

Do you mean that when there is a sidebar there is no way to produce a web page that is not linked from the sidebar? Or, it would be rendered only if we put `hidden: true` in its front matter? I like the idea of making it explicit to create those pages with a required property. Does the process fail if a page is present in the file system but not in the sidebar? Or is this file silently ignored? (BTW, I would call the property `orphan` instead of `hidden`, but YMMV)

pikinier20 commented:
> Does the process fail if a page is present in the file system but not in the sidebar? Or is this file silently ignored? (BTW, I would call the property `orphan` instead of `hidden`, but YMMV)

Currently, it silently ignores the file - actually, it just never checks any file that is not declared in YAML. We can change that to scan the directory and warn about files that exist but are not present in config

julienrf commented on scaladoc/src/dotty/tools/scaladoc/site/SidebarParser.scala:
```diff
@@ -6,31 +6,76 @@ import com.fasterxml.jackson.databind.ObjectMapper;
-    case RawInput(title, page, index, subsection) if page.isEmpty && !subsection.isEmpty() =>
-      Sidebar.Category(Option.when(title.nonEmpty)(title), Option.when(index.nonEmpty)(index), subsection.asScala.map(toSidebar).toList)
+  case class RawInput(
+    @BeanProperty var title: String,
```
Why do we need `@BeanProperty`?

pikinier20 replied:
Previously, we had explicitly defined setters. I found the `BeanProperty` shorter and resistant to possible errors (changed types or sth). The reason is that we use Jackson library that can only deserialize YAML to bean class.

julienrf commented on scaladoc/src/dotty/tools/scaladoc/site/SidebarParser.scala:
```diff
@@ -6,31 +6,76 @@ import com.fasterxml.jackson.databind.ObjectMapper;
+      Sidebar.Page(None, page, hidden)
+
+  private def schemaMessage: String =
+    s"""Static site YAML configuration file should comply to following model:
+      |case class Root(rootIndex: String, pages: List[Child])
+      |case class Page(title: Option[String], page: String) extends Child
+      |case class Subsection(
+      |  index: Option[String],
+      |  title: Option[String],
+      |  directory: Option[String],
+      |  subsection: List[Child]
+      |)
+      |""".stripMargin
```
I am not sure we should show a Scala type hierarchy to document a YAML schema. Maybe we should put a link to a documentation web page if it’s not possible to simply describe the YAML format in a message? Also, this type hierarchy does not compile (the type `Child` is not defined).

pikinier20 replied:
Ahh, I just didn't know how to define the format properly. I tried with JSON schema but it's too long. Maybe I'll try to just describe it in a less formal way.

julienrf commented on scaladoc/src/dotty/tools/scaladoc/site/common.scala:
```diff
@@ -58,7 +58,7 @@ final val LineSeparator = "\n"
 def yamlParser(using ctx: StaticSiteContext): Parser = Parser.builder(defaultMarkdownOptions).build()
 
-def loadTemplateFile(file: File)(using ctx: StaticSiteContext): TemplateFile = {
+def loadTemplateFile(file: File, titleOverride: Option[TemplateName] = None)(using ctx: StaticSiteContext): TemplateFile = {
```
Not very important but I don’t think `titleOverride` is a good name because it does not override a previously set title, instead it provides a default title in case it is unset. I suggest using `defaultTitle` instead.

julienrf commented:
Reading again the specification:

> We can also define a section with title and index but without defining children which will cause loading documents from the directory of index file

Does that overlap with the `hidden` property? It seems that it renders all the pages that are under the given `index` path, but without showing them in the side bar?

julienrf commented:
I think this is going in a good direction, could you add some user documentation? Most probably the best place to do so for now would be here: https://github.com/scala/docs.scala-lang/tree/main/_overviews/scala3-scaladoc

pikinier20 commented:
> Reading again the specification:
> 
> > We can also define a section with title and index but without defining children which will cause loading documents from the directory of index file
> 
> Does that overlap with the `hidden` property? It seems that it renders all the pages that are under the given `index` path, but without showing them in the side bar?

If we specify a section without children, we get all pages that are in the directory and we add them all to navbar. - It's a shortcut to people that just want to have all pages without writing long YAML files.

pikinier20 commented on scaladoc/src/dotty/tools/scaladoc/site/SidebarParser.scala:
```diff
@@ -6,31 +6,86 @@ import com.fasterxml.jackson.databind.ObjectMapper;
+
+  private def schemaMessage: String =
+    s"""Static site YAML configuration file should comply to the following description:
+      |rootIndex: <string> # optional
+      |pages:
+      |  - <subsection> | <page>
+      |
+      |<subsection>:
+      |  title: <string> # optional
+      |  index: <string> # optional
+      |  directory: <string> # optional
+      |  subsection: # optional
+      |    - <subsection> | <page>
+      |  # either index or subsection needs to be present
+      |<page>:
+      |  title: <string> # optional
+      |  page: <string>
+      |  hidden: <boolean> # optional
+      |
+      |For more information visit:
+      |https://docs.scala-lang.org/scala3/guides/scaladoc/static-site.html
```
Uniformizing sounds reasonable

pikinier20 commented:
@julienrf I think I addressed all issues that you pointed out. We need there @BarkingBad to re-review the PR. In the meantime we can take a look at the entire work and apply some polishing if there's a need.

julienrf reviewed (approved):
Nice!
Just to double-check, so how does the landing page look like now?

pikinier20 commented:
> Nice! Just to double-check, so how does the landing page look like now?

Right now, it's the main page of docs.

Nicolapps commented:
It looks to me that many URLs are broken by this change. For instance, https://dotty.epfl.ch/docs/usage/scaladoc/ is now located at https://dotty.epfl.ch/usage/scaladoc/.

Would it be possible to put redirections in place?

Also, it might be necessary to update links in the official Scala repositories (I can count 80 links in [docs.scala-lang](https://github.com/scala/docs.scala-lang) and 214 links in [Dotty](https://github.com/lampepfl/dotty) itself).

Nicolapps commented:
@julienrf Thank you for your reply. So, what should be done in the meantime for all the broken links in the official Scala repositories? Would a pull request fixing them be welcome?

julienrf commented:
Hey Nicolas, that’s really nice of you to propose your help!
I think that should not be necessary because there is an ongoing PR to restore the old URLs: https://github.com/lampepfl/dotty/pull/14475
